core/src/main/scala/org/apache/spark/scheduler/Pool.scala

```diff
@@ -119,4 +119,66 @@ private[spark] class Pool(
       parent.decreaseRunningTasks(taskNum)
     }
   }
+
+  // Update the number of slots considered available for each TaskSetManager whose ancestor
+  // in the tree is this pool
+  // For FAIR scheduling, slots are distributed among pools based on weights and minshare.
+  //   If a pool requires fewer slots than are available to it, the leftover slots are redistributed
+  //   to the remaining pools using the remaining pools' weights.
+  // For FIFO scheduling, the schedulable queue is iterated over in FIFO order,
+  //   giving each schedulable the remaining slots,
+  //   up to the number of remaining tasks for that schedulable.
+  override def updateAvailableSlots(numSlots: Float): Unit = {
+    schedulingMode match {
+      case SchedulingMode.FAIR =>
+        val queueCopy = new ConcurrentLinkedQueue[Schedulable](schedulableQueue)
+        var shouldRedistribute = false
+        var totalWeights = schedulableQueue.asScala.map(_.weight).sum
+        var totalSlots = numSlots
+        do {
+          shouldRedistribute = false
+          var nextWeights = totalWeights
+          var nextSlots = totalSlots
+          queueCopy.forEach(schedulable => {
+            val numTasksRemaining = schedulable.getSortedTaskSetQueue
+              .map(tsm => tsm.tasks.length - tsm.tasksSuccessful).sum
+            val allocatedSlots = Math.max(
+              totalSlots * schedulable.weight / totalWeights,
+              schedulable.minShare)
+            if (numTasksRemaining < allocatedSlots) {
+              schedulable.updateAvailableSlots(numTasksRemaining)
+              nextWeights -= schedulable.weight
+              nextSlots -= numTasksRemaining
+              shouldRedistribute = true
+              queueCopy.remove(schedulable)
+            }
+          })
+          totalWeights = nextWeights
+          totalSlots = nextSlots
+        } while (shouldRedistribute)
+
+        // All schedulables remaining have more remaining tasks than their share of slots,
+        // so no need to recalculate remaining tasks. Just give them their total share of slots.
+        queueCopy.forEach(schedulable => {
+          val allocatedSlots = Math.max(
+            totalSlots * schedulable.weight / totalWeights,
+            schedulable.minShare)
+          schedulable.updateAvailableSlots(allocatedSlots)
+        })
+      case SchedulingMode.FIFO =>
+        val sortedSchedulableQueue =
+          schedulableQueue.asScala.toSeq.sortWith(taskSetSchedulingAlgorithm.comparator)
+        var remainingSlots = numSlots
+        for (schedulable <- sortedSchedulableQueue) {
```

Comment:
we can stop the loop earlier if `remainingSlots` becomes 0.

Reply:
we shouldn't skip the loop because we need to update the remaining schedulables with 0 slots.
I've added a shortcut for the 0 case.